File: windows/RNTrackPlayer/Logic/MediaManager.cs
using Microsoft.ReactNative.Managed;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using TrackPlayer.Players;

namespace TrackPlayer.Logic
{
    public class MediaManager
    {
        private TrackPlayerModule module;

        private Metadata metadata;

        private Playback player;

        public MediaManager(TrackPlayerModule module)
        {
            this.module = module;
            this.metadata = new Metadata(this, module);
        }

        public void SwitchPlayback(Playback pb)
        {
            if (player != null)
            {
                player.Stop();
                player.Dispose();
            }

            player = pb;
            metadata.SetTransportControls(pb?.GetTransportControls());
        }

        public LocalPlayback CreateLocalPlayback(JSValueObject options)
        {
            return new LocalPlayback(this, options);
        }

        public void UpdateOptions(JSValue options)
        {
            metadata.UpdateOptions(options);
        }

        public Playback GetPlayer()
        {
            return player;
        }

        public Metadata GetMetadata()
        {
            return metadata;
        }

        public void OnEnd(Track previous, double prevPos)
        {
            module.PlaybackQueueEnded(new JSValueObject{ {"track", previous?.Id },
                                                         {"position", prevPos } });
        }

        public void OnStateChange(PlaybackState state)
        {
            module.PlaybackStateAction(new JSValueObject { { "state", (int)state } });
        }

        public void OnTrackUpdate(Track previous, double prevPos, Track next, bool changed)
        {
            metadata.UpdateMetadata(next);

            if (changed)
            {
                var jvo = new JSValueObject{{ "track", previous?.Id },
                                            { "position", prevPos },
                                            { "nextTrack", next?.Id } };
                module.PlaybackTrackChanged(jvo);
            }
        }

        public void OnError(string code, string error)
        {
            JSValueObject jvo = new JSValueObject{{ "code", code },
                                                  { "message", error } };
            module.PlaybackError(jvo);
        }

        public void Dispose()
        {
            if (player != null)
            {
                player.Dispose();
                player = null;
            }

            metadata.Dispose();
        }
    }
}

File: windows/RNTrackPlayer/Logic/MediaManager.cpp
#include "pch.h"
#include "Logic/MediaManager.h"
#include "Logic/Events.h"
#include "Logic/Track.h"

using namespace winrt::RNTrackPlayer;
using namespace winrt::Windows::UI::Xaml;

MediaManager::MediaManager(React::ReactContext const& context)
    : player(nullptr),
      metadata(*this)
{
    this->context = context;
}

void MediaManager::SendEvent(const std::string& eventName, const JSValueObject& data)
{
    std::wstring _eventName(eventName.begin(), eventName.end());
    context.EmitJSEvent(L"RCTDeviceEventEmitter", _eventName, data);
}

void MediaManager::SwitchPlayback(Playback* pb)
{
    if (player != nullptr)
    {
        player->Stop();
        delete player;
    }

    player = pb;
    if (pb)
    {
        metadata.SetTransportControls(pb->GetTransportControls());
    }
}

LocalPlayback* MediaManager::CreateLocalPlayback(React::JSValueObject& options)
{
    return new LocalPlayback(*this, options);
}

void MediaManager::UpdateOptions(React::JSValueObject& options)
{
    metadata.UpdateOptions(options);
}

Playback* MediaManager::GetPlayer() const
{
    return const_cast<Playback*>(player);
}

Metadata* MediaManager::GetMetadata() const
{
    return const_cast<Metadata*>(&metadata);
}

void MediaManager::OnEnd(int previous, double prevPos)
{
    VERBOSE_DEBUG("OnEnd");

    JSValueObject obj;
    if (previous != -1)
    {
        obj["track"] = previous;
    }
    obj["position"] = prevPos;
    SendEvent(Events::PlaybackQueueEnded, obj);
}

void MediaManager::OnStateChange(PlaybackState state)
{
    VERBOSE_DEBUG("OnStateChange");

    JSValueObject obj;
    obj["state"] = (int)state;
    SendEvent(Events::PlaybackState, obj);
}

void MediaManager::OnTrackUpdate(int previousIndex, double prevPos, int nextIndex, Track* next)
{
    VERBOSE_DEBUG("OnTrackUpdate");

    if (next)
    {
        metadata.UpdateMetadata(*next);
    }

    JSValueObject obj;
    if (previousIndex != -1)
    {
        obj["track"] = previousIndex;
    }
    obj["position"] = prevPos;
    if (nextIndex != -1)
    {
        obj["nextTrack"] = nextIndex;
    }

    SendEvent(Events::PlaybackTrackChanged, obj);
}

void MediaManager::OnError(const std::string& code, const std::string& error)
{
    VERBOSE_DEBUG(error);

    JSValueObject obj;
    obj["code"] = code;
    obj["message"] = error;
    SendEvent(Events::PlaybackError, obj);
}

